Move WebExtensionMatchPattern to cross-platform location.

xeenon opened the pull request:
#### 7d1d793c1d1794f1792401ae3f00583163457b3a
<pre>
Move WebExtensionMatchPattern to cross-platform location.
<a href="https://webkit.org/b/279734">https://webkit.org/b/279734</a>
<a href="https://rdar.apple.com/problem/136037431">rdar://problem/136037431</a>

Reviewed by Brian Weinstein.

Moved the bulk of the implementation as-is from `WebExtensionMatchPatternCocoa`. Changed
the errors to be localized, since `API::Error` requires it. Also changed `expandedStrings()`
to return a `Vector`.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toErrorString): Move `trimTrailingPeriod()` to the result of the format, so it catches
trailing periods in trailing substitution strings.
* Source/WebKit/Sources.txt: Added WebExtensionMatchPattern.cpp.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionMatchPattern.mm:
(-[WKWebExtensionMatchPattern initWithString:error:]):
(-[WKWebExtensionMatchPattern initWithScheme:host:path:error:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::InjectedContentData::expandedIncludeMatchPatternStrings const):
(WebKit::WebExtension::InjectedContentData::expandedExcludeMatchPatternStrings const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::corsDisablingPatterns):
(WebKit::WebExtensionContext::addInjectedContent):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::updateWebsitePoliciesForNavigation):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::toPatterns):
(WebKit::toAPI):
(WebKit::WebExtensionMatchPattern::extensionSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::validSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::supportedSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchURL): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchPattern): Deleted.
(WebKit::patternCache): Deleted.
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme): Deleted.
(WebKit::WebExtensionMatchPattern::isWebExtensionURL): Deleted.
(WebKit::WebExtensionMatchPattern::getOrCreate): Deleted.
(WebKit::WebExtensionMatchPattern::allURLsMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::allHostsAndSchemesMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts): Deleted.
(WebKit::error): Deleted.
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::isSupported const): Deleted.
(WebKit::WebExtensionMatchPattern::operator== const): Deleted.
(WebKit::WebExtensionMatchPattern::scheme const): Deleted.
(WebKit::WebExtensionMatchPattern::host const): Deleted.
(WebKit::WebExtensionMatchPattern::path const): Deleted.
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const): Deleted.
(WebKit::WebExtensionMatchPattern::stringWithScheme const): Deleted.
(WebKit::WebExtensionMatchPattern::expandedStrings const): Deleted.
(WebKit::WebExtensionMatchPattern::matchesAllHosts const): Deleted.
(WebKit::WebExtensionMatchPattern::isValidScheme): Deleted.
(WebKit::WebExtensionMatchPattern::matchesURL const): Deleted.
(WebKit::WebExtensionMatchPattern::matchesPattern const): Deleted.
(WebKit::WebExtensionMatchPattern::schemeMatches const): Deleted.
(WebKit::WebExtensionMatchPattern::hostMatches const): Deleted.
(WebKit::WebExtensionMatchPattern::pathMatches const): Deleted.
(WebKit::toStrings): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::matches const):
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.cpp: Copied from Cocoa/WebExtensionMatchPatternCocoa.mm.
(WebKit::WebExtensionMatchPattern::extensionSchemes):
(WebKit::WebExtensionMatchPattern::validSchemes):
(WebKit::WebExtensionMatchPattern::supportedSchemes):
(WebKit::WebExtensionMatchPattern::patternsMatchURL):
(WebKit::WebExtensionMatchPattern::patternsMatchPattern):
(WebKit::patternCache):
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme):
(WebKit::WebExtensionMatchPattern::isWebExtensionURL):
(WebKit::WebExtensionMatchPattern::getOrCreate):
(WebKit::WebExtensionMatchPattern::allURLsMatchPattern):
(WebKit::WebExtensionMatchPattern::allHostsAndSchemesMatchPattern):
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts):
(WebKit::WebExtensionMatchPattern::createError):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::operator== const):
(WebKit::WebExtensionMatchPattern::scheme const):
(WebKit::WebExtensionMatchPattern::host const):
(WebKit::WebExtensionMatchPattern::path const):
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const):
(WebKit::WebExtensionMatchPattern::stringWithScheme const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::matchesAllHosts const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL const):
(WebKit::WebExtensionMatchPattern::matchesPattern const):
(WebKit::WebExtensionMatchPattern::schemeMatches const):
(WebKit::WebExtensionMatchPattern::hostMatches const):
(WebKit::WebExtensionMatchPattern::pathMatches const):
(WebKit::toStrings):
(WebKit::toPatterns):
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::wrapper const):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::~WebExtensionMatchPattern): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added WebExtensionMatchPattern.cpp and WebExtension.cpp.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionMatchPattern.mm:
(TestWebKitAPI::TEST(WKWebExtensionMatchPattern, PatternValidity)): Use `error.localizedDescription`.
(TestWebKitAPI::TEST(WKWebExtensionMatchPattern, CustomURLScheme)): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, Initialization)): Remove double period after fix in `toErrorString()`.

Canonical link: <a href="https://commits.webkit.org/283710@main">https://commits.webkit.org/283710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9af6fd51f583e177abef17829615f8860ab9eb8c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67160 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/46539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19792 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/18293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69278 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/54337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/18086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/71195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/18293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70227 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/54337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/58113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/54337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/15507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/16647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/54337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/15848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/11119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/18086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/72898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/11152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/58172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/2739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10190 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/43421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/44607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/43162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -55,6 +56,14 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
+        InvalidScheme,
+        InvalidHost,
+        InvalidPath,
+    };
```
Will we continue to keep both of these around? It would be nice to not need this duplication in the future.

xeenon replied:
One is public ObjC API, which can't be included here, and this is internal which is typical for WebKit to have both.

xeenon committed:
Move WebExtensionMatchPattern to cross-platform location.
https://webkit.org/b/279734
rdar://problem/136037431

Reviewed by Brian Weinstein.

Moved the bulk of the implementation as-is from `WebExtensionMatchPatternCocoa`. Changed
the errors to be localized, since `API::Error` requires it. Also changed `expandedStrings()`
to return a `Vector`.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::toErrorString): Move `trimTrailingPeriod()` to the result of the format, so it catches
trailing periods in trailing substitution strings.
* Source/WebKit/Sources.txt: Added WebExtensionMatchPattern.cpp.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionMatchPattern.mm:
(-[WKWebExtensionMatchPattern initWithString:error:]):
(-[WKWebExtensionMatchPattern initWithScheme:host:path:error:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::InjectedContentData::expandedIncludeMatchPatternStrings const):
(WebKit::WebExtension::InjectedContentData::expandedExcludeMatchPatternStrings const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::corsDisablingPatterns):
(WebKit::WebExtensionContext::addInjectedContent):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::updateWebsitePoliciesForNavigation):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::toPatterns):
(WebKit::toAPI):
(WebKit::WebExtensionMatchPattern::extensionSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::validSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::supportedSchemes): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchURL): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchPattern): Deleted.
(WebKit::patternCache): Deleted.
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme): Deleted.
(WebKit::WebExtensionMatchPattern::isWebExtensionURL): Deleted.
(WebKit::WebExtensionMatchPattern::getOrCreate): Deleted.
(WebKit::WebExtensionMatchPattern::allURLsMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::allHostsAndSchemesMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts): Deleted.
(WebKit::error): Deleted.
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::isSupported const): Deleted.
(WebKit::WebExtensionMatchPattern::operator== const): Deleted.
(WebKit::WebExtensionMatchPattern::scheme const): Deleted.
(WebKit::WebExtensionMatchPattern::host const): Deleted.
(WebKit::WebExtensionMatchPattern::path const): Deleted.
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const): Deleted.
(WebKit::WebExtensionMatchPattern::stringWithScheme const): Deleted.
(WebKit::WebExtensionMatchPattern::expandedStrings const): Deleted.
(WebKit::WebExtensionMatchPattern::matchesAllHosts const): Deleted.
(WebKit::WebExtensionMatchPattern::isValidScheme): Deleted.
(WebKit::WebExtensionMatchPattern::matchesURL const): Deleted.
(WebKit::WebExtensionMatchPattern::matchesPattern const): Deleted.
(WebKit::WebExtensionMatchPattern::schemeMatches const): Deleted.
(WebKit::WebExtensionMatchPattern::hostMatches const): Deleted.
(WebKit::WebExtensionMatchPattern::pathMatches const): Deleted.
(WebKit::toStrings): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::matches const):
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.cpp: Copied from Cocoa/WebExtensionMatchPatternCocoa.mm.
(WebKit::WebExtensionMatchPattern::extensionSchemes):
(WebKit::WebExtensionMatchPattern::validSchemes):
(WebKit::WebExtensionMatchPattern::supportedSchemes):
(WebKit::WebExtensionMatchPattern::patternsMatchURL):
(WebKit::WebExtensionMatchPattern::patternsMatchPattern):
(WebKit::patternCache):
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme):
(WebKit::WebExtensionMatchPattern::isWebExtensionURL):
(WebKit::WebExtensionMatchPattern::getOrCreate):
(WebKit::WebExtensionMatchPattern::allURLsMatchPattern):
(WebKit::WebExtensionMatchPattern::allHostsAndSchemesMatchPattern):
(WebKit::WebExtensionMatchPattern::patternsMatchAllHosts):
(WebKit::WebExtensionMatchPattern::createError):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::operator== const):
(WebKit::WebExtensionMatchPattern::scheme const):
(WebKit::WebExtensionMatchPattern::host const):
(WebKit::WebExtensionMatchPattern::path const):
(WebKit::WebExtensionMatchPattern::hostIsPublicSuffix const):
(WebKit::WebExtensionMatchPattern::stringWithScheme const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::matchesAllHosts const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL const):
(WebKit::WebExtensionMatchPattern::matchesPattern const):
(WebKit::WebExtensionMatchPattern::schemeMatches const):
(WebKit::WebExtensionMatchPattern::hostMatches const):
(WebKit::WebExtensionMatchPattern::pathMatches const):
(WebKit::toStrings):
(WebKit::toPatterns):
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
(WebKit::WebExtensionMatchPattern::wrapper const):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern): Deleted.
(WebKit::WebExtensionMatchPattern::~WebExtensionMatchPattern): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added WebExtensionMatchPattern.cpp and WebExtension.cpp.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionMatchPattern.mm:
(TestWebKitAPI::TEST(WKWebExtensionMatchPattern, PatternValidity)): Use `error.localizedDescription`.
(TestWebKitAPI::TEST(WKWebExtensionMatchPattern, CustomURLScheme)): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIWebRequest, Initialization)): Remove double period after fix in `toErrorString()`.

Canonical link: https://commits.webkit.org/283710@main
